### DISCUSSION TITLE
Static cache how to exclude class

### DISCUSSION THREAD

**Discussion by earthwytch** [statamic/cms#8553] [Q&A] [answered]
In the `static_caching.php` config file there is a tantalising option for "class" but I can't find anything in the docs about this and Google has failed me.

```
    'exclude' => [

        'class' => null,

        'urls' => [
            '/search',
            '/contact-us'
        ],

    ],
```

Is this an option I could use to exclude all pages with forms on them, so I don't have to individually add each form url?

**Statamic v4.16.0**

**Comment by jasonvarga** [comment, marked as answer]
You can create a class that implements `Statamic\StaticCaching\UrlExcluder` then point to it there.

```php
'exclude' => [
    'class' => \App\MyUrlExcluder::class,
],
```

```php
<?php

namespace App;

use Statamic\StaticCaching\UrlExcluder;

class MyUrlExcluder implements UrlExcluder
{
    public function isExcluded(string $url): bool
    {
        // Your logic to figure out whether the url should be excluded.
    	return false;    
    }
}
```

That's how you'd use a custom class. However, I'm not sure how you'd be able to magically check "if the page has a form on it". 🤔

**Reply by earthwytch** [reply]
Thanks, that makes sense 👍 The main use case is if a user moves the un-cached page around in the tree it will break the page. This is likely to feel really random for them because they won't understand why. This might then make them feel kinda nervous about breaking things and less reluctant to do things, and I'd rather they feel reasonably empowered to make those sorts of changes themselves.

Since the forms are all using a specific page template and specific blueprints, I wonder if I could use that to have the urls automatically update to a list somewhere when the page is published, and then use MyUrlExcluder::class to query that list? Or perhaps easier, just add a "Cache page" toggle to all pages and have that update the list. I'll give it a go 🧙‍♀️

**Reply by jasonvarga** [reply]
You don't really need to exclude form pages anymore. CSRF tokens are automatically updated on cached pages. You really just need it if there's dynamic stuff in it like conditional success messages.

You could consider just using the nocache tag around your forms. https://statamic.dev/tags/nocache#forms

**Reply by earthwytch** [reply]
Thanks 👍 the trouble with the nocache tag is the javascript it inserts which I haven't been able to find a way to add a nonce to for CSP yet (another discussion). So I'm avoiding it where possible until I find a solution for that.
